File: src/lib/api/knowledge-domain/analysis.ts
import { devApiFetch, hasTauriRuntime, invoke } from "../client";
import { knowledgeApi } from "../knowledge";
import type {
  AnalyzeKnowledgeAnalysisSnapshotInput,
  CaptureKnowledgeAnalysisGitSnapshotInput,
  GenerateKnowledgeAnalysisDocumentsInput,
  CreateKnowledgeAnalysisDraftInput,
  ConfirmKnowledgeAnalysisDraftInput,
  ConfirmKnowledgeAnalysisDraftResult,
  GenerateKnowledgeCodeDocumentsResult,
  KnowledgeAnalysisDraft,
  KnowledgeCodeAnalysisResult,
  KnowledgeCodeSnapshot,
  KnowledgeCodeSource,
  ListKnowledgeAnalysisCodeSnapshotsInput,
  UpsertKnowledgeCodeSourceInput,
} from "@/types/knowledge-domain/analysis";

/**
 * 源码分析领域的真实 IPC 出口。桌面端使用新领域 Command；浏览器开发验收复用旧 Dev API
 * 的等价只读/固定报告路径；AI 草稿也通过同一项目版本、快照与来源授权边界。
 */
export const knowledgeAnalysisApi = {
  // 目录工作台首次登记的 Git 来源，同时登记为本地静态分析来源；调用仍复用旧服务，
  // 以保留路径、符号链接和远程处理授权的统一校验。
  upsertCodeSource: (input: UpsertKnowledgeCodeSourceInput) =>
    knowledgeApi.upsertCodeSource(input),
  listCodeSources: (projectId: number) =>
    hasTauriRuntime()
      ? invoke<KnowledgeCodeSource[]>("list_knowledge_analysis_code_sources", {
          projectId,
        })
      : devApiFetch<KnowledgeCodeSource[]>(
          `/knowledge/projects/${projectId}/analysis/code-sources`,
        ),
  listCodeSnapshots: ({
    projectId,
    sourceId,
  }: ListKnowledgeAnalysisCodeSnapshotsInput) =>
    hasTauriRuntime()
      ? invoke<KnowledgeCodeSnapshot[]>(
          "list_knowledge_analysis_code_snapshots",
          {
            projectId,
            sourceId,
          },
        )
      : devApiFetch<KnowledgeCodeSnapshot[]>(
          sourceId == null
            ? `/knowledge/projects/${projectId}/analysis/code-snapshots`
            : `/knowledge/projects/${projectId}/analysis/code-snapshots?sourceId=${sourceId}`,
        ),
  captureGitSnapshot: (input: CaptureKnowledgeAnalysisGitSnapshotInput) =>
    hasTauriRuntime()
      ? invoke<KnowledgeCodeSnapshot>(
          "capture_knowledge_analysis_git_snapshot",
          {
            projectId: input.projectId,
            input: {
              sourceId: input.sourceId,
              gitRef: input.gitRef,
              releaseId: input.projectVersionId ?? null,
            },
          },
        )
      : devApiFetch<KnowledgeCodeSnapshot>(
          `/knowledge/projects/${input.projectId}/analysis/code-snapshots/git`,
          {
            method: "POST",
            body: JSON.stringify({
              sourceId: input.sourceId,
              gitRef: input.gitRef,
              releaseId: input.projectVersionId ?? null,
            }),
          },
        ),
  analyzeSnapshot: ({
    projectId,
    snapshotId,
  }: AnalyzeKnowledgeAnalysisSnapshotInput) =>
    hasTauriRuntime()
      ? invoke<KnowledgeCodeAnalysisResult>(
          "analyze_knowledge_analysis_snapshot",
          {
            projectId,
            snapshotId,
          },
        )
      : devApiFetch<KnowledgeCodeAnalysisResult>(
          `/knowledge/projects/${projectId}/analysis/code-snapshots/${snapshotId}/analyze`,
          { method: "POST" },
        ),
  generateDocuments: ({
    projectId,
    snapshotId,
  }: GenerateKnowledgeAnalysisDocumentsInput) =>
    hasTauriRuntime()
      ? invoke<GenerateKnowledgeCodeDocumentsResult>(
          "generate_knowledge_analysis_documents",
          { projectId, input: { snapshotId } },
        )
      : devApiFetch<GenerateKnowledgeCodeDocumentsResult>(
          `/knowledge/projects/${projectId}/analysis/code-snapshots/documents/generate`,
          { method: "POST", body: JSON.stringify({ snapshotId }) },
        ),
  createAiDraft: (input: CreateKnowledgeAnalysisDraftInput) =>
    hasTauriRuntime()
      ? invoke<KnowledgeAnalysisDraft>("create_knowledge_analysis_ai_draft", {
          input,
        })
      : devApiFetch<KnowledgeAnalysisDraft>(
          `/knowledge/projects/${input.projectId}/analysis/ai-drafts`,
          { method: "POST", body: JSON.stringify(input) },
        ),
  confirmAiDraft: (input: ConfirmKnowledgeAnalysisDraftInput) =>
    hasTauriRuntime()
      ? invoke<ConfirmKnowledgeAnalysisDraftResult>(
          "confirm_knowledge_analysis_ai_draft",
          { input },
        )
      : devApiFetch<ConfirmKnowledgeAnalysisDraftResult>(
          `/knowledge/analysis/ai-drafts/${input.draftId}/confirm`,
          { method: "POST", body: JSON.stringify(input) },
        ),
};

File: src-tauri/src/commands/knowledge_domain/analysis.rs
use crate::error::CommandError;
use crate::models::knowledge_domain::analysis::{
    ConfirmKnowledgeAnalysisDraftInput, ConfirmKnowledgeAnalysisDraftResult,
    CreateKnowledgeAnalysisDraftInput, KnowledgeAnalysisDraft,
};
use crate::models::{
    CaptureKnowledgeGitSnapshotInput, GenerateKnowledgeCodeDocumentsInput,
    GenerateKnowledgeCodeDocumentsResult, KnowledgeCodeAnalysisResult, KnowledgeCodeSnapshot,
    KnowledgeCodeSource,
};
use crate::services::knowledge_domain::analysis::KnowledgeAnalysisService;
use crate::state::AppState;

pub(crate) const DOMAIN: &str = "analysis";

/// 仅列出当前项目登记的源码知识源，避免新工作台跨项目拼接源 ID。
#[tauri::command]
pub fn list_knowledge_analysis_code_sources(
    state: tauri::State<'_, AppState>,
    project_id: i64,
) -> Result<Vec<KnowledgeCodeSource>, CommandError> {
    KnowledgeAnalysisService::list_project_code_sources(&state.db, project_id).map_err(Into::into)
}

/// 仅列出当前项目的代码快照；指定 source 时仍会校验它属于该项目。
#[tauri::command]
pub fn list_knowledge_analysis_code_snapshots(
    state: tauri::State<'_, AppState>,
    project_id: i64,
    source_id: Option<i64>,
) -> Result<Vec<KnowledgeCodeSnapshot>, CommandError> {
    KnowledgeAnalysisService::list_project_code_snapshots(&state.db, project_id, source_id)
        .map_err(Into::into)
}

/// 捕获不可变 Git Commit 快照。输入仍复用旧 DTO，保证 releaseId、Git ref 等已验证字段
/// 不会在迁移期间漂移；项目 ID 是新工作台额外的授权范围。
#[tauri::command]
pub async fn capture_knowledge_analysis_git_snapshot(
    state: tauri::State<'_, AppState>,
    project_id: i64,
    input: CaptureKnowledgeGitSnapshotInput,
) -> Result<KnowledgeCodeSnapshot, CommandError> {
    KnowledgeAnalysisService::capture_git_snapshot(&state.db, project_id, input)
        .await
        .map_err(Into::into)
}

/// 运行现有的确定性静态分析；分析服务会维持快照状态和失败边界。
#[tauri::command]
pub async fn analyze_knowledge_analysis_snapshot(
    state: tauri::State<'_, AppState>,
    project_id: i64,
    snapshot_id: i64,
) -> Result<KnowledgeCodeAnalysisResult, CommandError> {
    KnowledgeAnalysisService::analyze_snapshot(&state.db, project_id, snapshot_id)
        .await
        .map_err(Into::into)
}

/// 生成基于已分析快照的固定模板报告。该 Command 不生成、保存或确认 AI 草稿。
#[tauri::command]
pub fn generate_knowledge_analysis_documents(
    state: tauri::State<'_, AppState>,
    project_id: i64,
    input: GenerateKnowledgeCodeDocumentsInput,
) -> Result<GenerateKnowledgeCodeDocumentsResult, CommandError> {
    KnowledgeAnalysisService::generate_documents(&state.db, project_id, input).map_err(Into::into)
}

/// 仅能以同一项目版本中已冻结、已分析的 Git Commit 为证据生成 AI 草稿。正式文档写入
/// 仍须经过单独的人工确认 Command，避免远程模型输出直接进入知识库。
#[tauri::command]
pub async fn create_knowledge_analysis_ai_draft(
    state: tauri::State<'_, AppState>,
    input: CreateKnowledgeAnalysisDraftInput,
) -> Result<KnowledgeAnalysisDraft, CommandError> {
    KnowledgeAnalysisService::create_ai_draft(&state.db, input)
        .await
        .map_err(Into::into)
}

/// 用户编辑 AI 草稿后显式确认入库。服务层会创建新的知识文档版本并记录草稿与版本的
/// 审计关联，不覆盖历史文档或原始草稿。
#[tauri::command]
pub fn confirm_knowledge_analysis_ai_draft(
    state: tauri::State<'_, AppState>,
    input: ConfirmKnowledgeAnalysisDraftInput,
) -> Result<ConfirmKnowledgeAnalysisDraftResult, CommandError> {
    KnowledgeAnalysisService::confirm_ai_draft(&state.db, input).map_err(Into::into)
}
